File: Lista 2/GaloisField.py
import math
p = 1234567891

class GaloisField:
    def __init__(self, x=0):
        self.x = x
    #x + y
    def __add__(self, other):
        return GaloisField((self.x + other.x) % p)
    #x - y
    def __sub__(self, other):
        return GaloisField((self.x - other.x) % p)
    #x * y
    def __mul__(self, other):
        return GaloisField((self.x * other.x) % p)
    #x / y
    def __truediv__(self, other):
        g, x, y = gcdExtended(other.x, p)
        if (g != 1):
            inv = -1
        else:
            inv = (x%p + p) % p
        self.x = self.x % p
        if (inv == -1):
            print("Division not defined")
        else:
            return GaloisField((inv * self.x) % p)
    #x ^ y
    def __pow__(self, other):
        return GaloisField(pow(self.x, other.x, p))
    #x == y
    def __eq__(self, other):
        if (self.x%p == other.x%p):
            return True;
        else:
            return False;
    #x < y
    def __lt__(self, other):
        if (self.x%p < other.x%p):
            return True;
        else:
            return False;
    #x > y
    def __gt__ (self, other):
        if (self.x%p > other.x%p):
            return True;
        else:
            return False;
    #x <= y
    def __le__(self, other):
        if (self.x%p <= other.x%p):
            return True;
        else:
            return False;
    #x >= y
    def __ge__(self, other):
        if (self.x%p >= other.x%p):
            return True;
        else:
            return False;
    #x != y
    def __ne__(self, other):
        if (self.x%p != other.x%p):
            return True;
        else:
            return False;

    def __str__(self):
        return "{}".format(self.x)

def gcdExtended(a, b):
    if a == 0:
        return b, 0, 1
    gcd, x1, y1 = gcdExtended(b%a, a)
    x = y1 - (b//a) * x1
    y = x1
    return gcd, x, y


def main():
    while True:
       print("Enter an operator (+, -, *, /, ^, ==, <, >, <=, >=, !=): To quit press 'q'!")
       option = str(input().strip())
       print("Enter two numbers:")
       x = int(input().strip())
       y = int(input().strip())
       num1 = GaloisField(x)
       num2 = GaloisField(y)
       if option == '+':
            print('{} + {} = {}'.format(num1, num2, num1 + num2))      
       elif option == '-':
            print('{} - {} = {}'.format(num1, num2, num1 - num2)) 
       elif option == '*':
            print('{} * {} = {}'.format(num1, num2, num1 * num2)) 
       elif option == '/':
            print('{} / {} = {}'.format(num1, num2, num1 / num2)) 
       elif option == '^':
            print('{} ^ {} = {}'.format(num1, num2, num1 ** num2)) 
       elif option == '==':
            print('{} == {}  {}'.format(num1, num2, num1 == num2)) 
       elif option == '<':
            print('{} < {}  {}'.format(num1, num2, num1 < num2)) 
       elif option == '>':
            print('{} > {}  {}'.format(num1, num2, num1 > num2)) 
       elif option == '<=':
            print('{} <= {}  {}'.format(num1, num2, num1 <= num2)) 
       elif option == '>':
            print('{} >= {}  {}'.format(num1, num2, num1 >= num2)) 
       elif option == '!=':
            print('{} != {}  {}'.format(num1, num2, num1 != num2)) 
       elif option == 'q':
            break
       else:
        continue

if __name__ == '__main__':
    main()

File: Lista 2/GaloisField.cpp
#include <stdio.h>
#include <iostream>
#include <math.h>

using namespace std;

const int p = 1234567891;

class GaloisField {

public:

    unsigned long long int x;

public:

    GaloisField() {
        this->x = 0;
    }

    GaloisField(int x) {
        this->x = x;
    }

    GaloisField operator+(const GaloisField & gf) {
        return GaloisField((this->x + gf.x)%p);
    }
    
    GaloisField operator -(const GaloisField & gf) {
        return GaloisField((this->x - gf.x)%p);
    }

    GaloisField operator *(const GaloisField & gf) {
        return GaloisField((this->x * gf.x)%p);
    }

    GaloisField operator /(const GaloisField & gf) {
        int x, y, inv;
        int g = extendedEuclideanAlgorithm(gf.x, p, &x, &y);
        if(g != 1)
            inv = -1;
        else    
            inv = (x%p + p) % p;
        this->x = this->x%p;
        if (inv == -1)
            return GaloisField(0);
        else
            return GaloisField((inv * this->x)%p);
    }

    GaloisField operator ^(const GaloisField & gf) {
        unsigned long long int result = 1;
        int y = gf.x;
        while (y > 0) {
            if (y % 2 == 1)
                result = (result * this->x);
            y = y >> 1;
            this->x = (this->x * this->x);
        }
        return GaloisField(result%p);
    }


    bool operator ==(const GaloisField & gf) {
        if (this->x%p == gf.x%p)
            return true;
        else
            return false;
    }  

    bool operator !=(const GaloisField & gf) {
        if (this->x%p != gf.x%p)
            return true;
        else
            return false;
    }  

    bool operator >(const GaloisField & gf) {
        if (this->x%p > gf.x%p)
            return true;
        else
            return false;
    }  

    bool operator <(const GaloisField & gf) {
        if (this->x%p < gf.x%p)
            return true;
        else
            return false;
    }  

    bool operator >=(const GaloisField & gf) {
        if (this->x%p >= gf.x%p)
            return true;
        else
            return false;
    }  

    bool operator <=(const GaloisField & gf) {
        if (this->x%p <= gf.x%p)
            return true;
        else
            return false;
    }  

    int extendedEuclideanAlgorithm(int a, int b, int *x, int *y) {
        if (a == 0){
            *x = 0, *y = 1;
            return b;
        }
        int x1, y1;
        int gcd = extendedEuclideanAlgorithm(b%a, a, &x1, &y1);
        *x = y1 - (b/a) *x1;
        *y = x1;
        return gcd;
    }


};

ostream& operator <<(ostream& s, const GaloisField& gf) {
    return s << gf.x;
}

void menu(int oper, GaloisField num1, GaloisField num2) {
    switch (oper) {
        case 1:
            cout << num1 << " + " << num2 << " = " << num1 + num2 << endl;
            break;
        case 2:
            cout << num1 << " - " << num2 << " = " << num1 - num2 << endl;
            break;
        case 3:
            cout << num1 << " * " << num2 << " = " << num1 * num2 << endl;
            break;
        case 4:
            cout << num1 << " / " << num2 << " = " << num1 / num2 << endl;
            break;
        case 5:
            cout << num1 << " ^ " << num2 << " = " << (num1 ^ num2) << endl;
            break;
        case 6:
            cout << num1 << " == " << num2 << "  " << (num1 == num2) << endl;
            break;
        case 7:
            cout << num1 << " < " << num2 << "  " << (num1 < num2) << endl;
            break;
        case 8:
            cout << num1 << " > " << num2 << "  " << (num1 > num2) << endl;
            break;
        case 9:
            cout << num1 << " <= " << num2 << "  " << (num1 <= num2) << endl;
            break;
        case 10:
            cout << num1 << " >= " << num2 << "  " << (num1 >= num2) << endl;
            break;
        case 11:
            cout << num1 << " != " << num2 << "  " << (num1 != num2) << endl;
            break;
        default:
            break;
        }
        return;
}

int main() {
    bool cond = true;
    int oper;
    GaloisField num1, num2;
    while (true) {
        cout << "Enter an operation\n 1. '+'\n 2. '-'\n 3. '*'\n 4. '/'\n 5. '^'\n 6. '=='\n 7. '<'\n 8. '>'\n 9. '<='\n 10. '>='\n 11. '!='\n To quit press '0'!\n ";
        cin >> oper;
        if (oper == 0)
            break;
        cout << "Enter two numbers: ";
        cin >> num1.x >> num2.x;
        menu(oper, num1, num2);
    }
    return 0;
}